src/v/utils/delta_for.h

```diff
@@ -88,18 +88,31 @@ struct delta_delta {
         auto p = last;
         uint64_t agg = 0;
         for (uint32_t i = 0; i < row_width; ++i) {
-            vassert(
-              row[i] >= p,
-              "Value {} can't be smaller than the previous one {}",
-              row[i],
-              p);
+            if (unlikely(row[i] < p)) {
+                ss::throw_with_backtrace<std::logic_error>(fmt::format(
+                  "Value row[{}]={} can't be smaller than the previous one "
+                  "{}. ",
+                  i,
+                  row[i],
+                  p));
+            }
+#ifndef NDEBUG
+            value_t delta;
+            if (unlikely(__builtin_sub_overflow(row[i], p, &delta))) {
+                ss::throw_with_backtrace<std::logic_error>(fmt::format(
+                  "overflow in op row[{}] - p -> {} {}", i, row[i], p));
+            }
```

Comment:
Is the extra check costly enough to warrant separation between debug and release builds? To my mind this is still very cheap and adding new fields to the encode is not on the hot path anyway.

Reply:
probably not, i was trying to keep the behavioral changes to a minimum.